From ffc313c67ba506c4719cef0685b5f0a4971c6ee9 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Timo=20Ter=C3=A4s?= <timo.teras@iki.fi>
Date: Tue, 2 Apr 2024 00:36:25 +0300
Subject: [PATCH] pkg: always use real digest, openssl-fips module crashes
 otherwise

Turns out the fips module crashes if EVP_md_null is used.

fixes #10992
---
 src/package.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/src/package.c b/src/package.c
index 8967389..05ebcf8 100644
--- a/src/package.c
+++ b/src/package.c
@@ -479,20 +479,19 @@ void apk_sign_ctx_init(struct apk_sign_ctx *ctx, int action,
 	case APK_SIGN_VERIFY:
 		/* If we're only verifing, we're going to start with a
 		 * signature section, which we don't need a hash of */
-		ctx->md = EVP_md_null();
 		ctx->verify_error = -ENOKEY;
 		break;
 	case APK_SIGN_VERIFY_IDENTITY:
 		/* If we're checking the package against a particular hash,
 		 * we need to start with that hash, because there may not
 		 * be a signature section to deduce it from */
-		ctx->md = EVP_sha1();
 		memcpy(&ctx->identity, identity, sizeof(ctx->identity));
 		break;
 	default:
 		assert(!"valid sign mode");
 		break;
 	}
+	ctx->md = EVP_sha1();
 	ctx->mdctx = EVP_MD_CTX_new();
 	EVP_DigestInit_ex(ctx->mdctx, ctx->md, NULL);
 }
-- 
2.44.0

